curp/src/client_new/mod.rs

```diff
@@ -0,0 +1,168 @@
+/// Unary rpc client
+mod unary;
+
+use std::{collections::HashMap, sync::Arc};
+
+use async_trait::async_trait;
+use curp_external_api::cmd::Command;
+use futures::{stream::FuturesUnordered, StreamExt};
+use utils::config::ClientConfig;
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        connect::ConnectApi, protocol_client::ProtocolClient, ConfChange, FetchClusterRequest,
+        FetchClusterResponse, Member, ReadState,
+    },
+};
+
+/// The response of propose command, deserialized from [`crate::rpc::ProposeResponse`] or
+/// [`crate::rpc::WaitSyncedResponse`].
+#[allow(type_alias_bounds)] // that's not bad
+type ProposeResponse<C: Command> = Result<(C::ER, Option<C::ASR>), C::Error>;
+
+/// `ClientApi`, a higher wrapper for `ConnectApi`, providing some methods for communicating to
+/// the whole curp cluster. Automatically discovery curp server to update it's quorum.
+#[async_trait]
+pub trait ClientApi<C: Command> {
+    /// The client error
+    type Error;
+
+    /// Get the local connection when the client is on the server node.
+    fn local_connect(&self) -> Option<Arc<dyn ConnectApi>>;
+
+    /// Send propose to the whole cluster, `use_fast_path` set to `false` to fallback into ordered
+    /// requests (event the requests are commutative).
+    async fn propose(
+        &self,
+        cmd: &C,
+        use_fast_path: bool,
+    ) -> Result<ProposeResponse<C>, Self::Error>;
+
+    /// Send propose configuration changes to the cluster
+    async fn propose_conf_change(
+        &self,
+        changes: Vec<ConfChange>,
+    ) -> Result<Vec<Member>, Self::Error>;
+
+    /// Send propose to shutdown cluster
+    async fn propose_shutdown(&self) -> Result<(), Self::Error>;
+
+    /// Send fetch read state from leader
+    async fn fetch_read_state(&self, cmd: &C) -> Result<ReadState, Self::Error>;
+
+    /// Send fetch cluster requests to all servers (That's because initially, we didn't
+    /// know who the leader is.)
+    /// Note: The fetched cluster may still be outdated if `linearizable` is false
+    async fn fetch_cluster(&self, linearizable: bool) -> Result<FetchClusterResponse, Self::Error>;
+
+    /// Fetch leader id
+    #[inline]
+    async fn fetch_leader_id(&self, linearizable: bool) -> Result<ServerId, Self::Error> {
+        if linearizable {
+            let resp = self.fetch_cluster(true).await?;
+            return Ok(resp.leader_id.unwrap_or_else(|| {
+                unreachable!("linearizable fetch cluster should return a leader id")
+            }));
+        }
+        let resp = self.fetch_cluster(false).await?;
+        if let Some(id) = resp.leader_id {
+            return Ok(id);
+        }
+        // fallback to linearizable fetch
+        self.fetch_leader_id(true).await
+    }
+}
+
+/// Client builder to build a client
+#[derive(Debug, Clone)]
+pub struct ClientBuilder {
```

Comment:
We can use `derive_builder` crate to simply this `ClientBuilder`.

Reply:
It seems not possible. we don't have a `struct Client` to put #[derive(derive_builder::Builder)] on it. We only have the trait `ClientApi`.
Also, there are some async functions that `ClientBuilder` may use which we cannot directly use the function derived from `derive_builder`.

This is how we used the `derive_builder`:

https://github.com/xline-kv/Xline/blob/47bcd9faae6c3d5157657e50783feaddd3b25e83/curp/src/client.rs#L34-L54

I don't think it is a good practice because of many `#[builder(setter(skip))]` on fields. 

BTW, the codes in `utils/src/config.rs` seems more necessary to use `derive_builder` for optimization.